### PR TITLE
Layer control on touch device

### LINKED ISSUE #549: It's not possible to reorder layers on touch devices
(also, the touch gesture would clash with the context menu in the layer control?)

### PR DESCRIPTION
**Pull request by yhy0217**
- [x] make labels in layer control draggable, close #549
- [x] fix the bug that the opacity setting is expanded when open the layer control by a tap/long press
- [x] prevent the layer context menu from showing up when open the layer control by a long press
- [x] create tests

### CODE REVIEW

**Comment by prushforth** [comment]
@yhy0217 The case of [re-ordering multiple extents](https://maps4html.org/experiments/multiple-extents/) within a layer isn't managed yet, I think?  I see you got a test to work, good job on that!

**Comment by yhy0217** [comment]
> Edit: yeah perhaps I was mistaken, seems to work if I reload the page.

I worked on reordering extent things on this branch so I believe it should work, but yeah I found if you are using the desktop Chrome dev tool to emulate the mobile/touch device, you have to reload the page before carrying out any reordering operations (it is the Chrome issue I believe)

**Comment by prushforth** [comment]
> I worked on reordering extent things on this branch so I believe it should work, but yeah I found if you are using the desktop Chrome dev tool to emulate the mobile/touch device, you have to reload the page before carrying out any reordering operations (it is the Chrome issue I believe)

Exactly what I noticed.  It is clear we haven't done a lot of work for mobile users!  Thanks.

File: src/mapml/control/LayerControl.js [hunk 1]
```diff
@@ -146,16 +146,25 @@ export var LayerControl = L.Control.Layers.extend({
       ) return this;
 
       L.DomUtil.removeClass(this._container, 'leaflet-control-layers-expanded');
+      if (e.originalEvent && e.originalEvent.pointerType === 'touch') {
```
Comment (prushforth) [inline src/mapml/control/LayerControl.js:149]: I think [operator chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) will work here

File: src/mapml/layers/MapMLLayer.js [hunk 2]
```diff
@@ -499,17 +499,21 @@ export var MapMLLayer = L.Layer.extend({
...
         extentItemNameSpan.extent = this._extent._mapExtents[i];
 
-        extent.onmousedown = (downEvent) => {
-          if(downEvent.target.tagName.toLowerCase() === "input" || downEvent.target.tagName.toLowerCase() === "select") return;
-          downEvent.preventDefault();
+        extent.ontouchstart = extent.onmousedown = (downEvent) => {
+          if((downEvent.target.parentElement.tagName.toLowerCase() === 'label' &&
+              downEvent.target.tagName.toLowerCase() !== 'input') ||
+              downEvent.target.tagName.toLowerCase() === 'label') {          
+          // downEvent.preventDefault();
```
Comment (prushforth) [inline src/mapml/layers/MapMLLayer.js:506]: We should remove comments if they're truly no longer required.

File: src/mapml/layers/MapMLLayer.js [hunk 3]
```diff
@@ -499,17 +499,21 @@ export var MapMLLayer = L.Layer.extend({
         extent.setAttribute('aria-labelledby', extentItemNameSpan.id);
         extentItemNameSpan.extent = this._extent._mapExtents[i];
 
-        extent.onmousedown = (downEvent) => {
-          if(downEvent.target.tagName.toLowerCase() === "input" || downEvent.target.tagName.toLowerCase() === "select") return;
-          downEvent.preventDefault();
+        extent.ontouchstart = extent.onmousedown = (downEvent) => {
+          if((downEvent.target.parentElement.tagName.toLowerCase() === 'label' &&
```
Comment (prushforth) [inline src/mapml/layers/MapMLLayer.js:503]: Please be consistent with spacing in code blocks

```js
if (condition) {
  // do something
} else {
  // something else
}
```

Merci

File: src/mapml/layers/MapMLLayer.js [hunk 4]
```diff
@@ -499,17 +499,21 @@ export var MapMLLayer = L.Layer.extend({
...
+          // downEvent.preventDefault();
           downEvent.stopPropagation();
+          downEvent = downEvent instanceof TouchEvent? downEvent.touches[0] : downEvent;
 
           let control = extent,
               controls = extent.parentNode,
               moving = false, yPos = downEvent.clientY;
 
-              document.body.onmousemove = (moveEvent) => {
+              document.body.ontouchmove = document.body.onmousemove = (moveEvent) => {
```
Comment (prushforth) [inline src/mapml/layers/MapMLLayer.js:514]: I didn't realize you could do assignments like this. Interesting and concise.

File: src/mapml/layers/MapMLLayer.js [hunk 5]
```diff
@@ -499,17 +499,21 @@ export var MapMLLayer = L.Layer.extend({
...
                 moveEvent.preventDefault();
+                moveEvent = moveEvent instanceof TouchEvent? moveEvent.touches[0] : moveEvent;
```
Comment (prushforth) [inline src/mapml/layers/MapMLLayer.js:516]: spacing around `?`

File: src/mapml/layers/MapMLLayer.js [hunk 6]
```diff
@@ -562,12 +566,12 @@ export var MapMLLayer = L.Layer.extend({
                   zIndex++;
                 }
                 controls.classList.remove("mapml-draggable");
-                document.body.onmousemove = document.body.onmouseup = null;
+                document.body.ontouchmove = document.body.onmousemove = document.body.ontouchend = document.body.onmouseup = null;
```
Comment (prushforth) [inline src/mapml/layers/MapMLLayer.js:569]: I am wondering if we should be reviewing everywhere that we are using mouse events, should we review if touch events are necessary??

File: src/mapml/layers/MapMLLayer.js [hunk 7]
```diff
@@ -675,15 +684,19 @@ export var MapMLLayer = L.Layer.extend({
...
         fieldset.setAttribute('aria-labelledby', layerItemName.id);
 
-        fieldset.onmousedown = (downEvent) => {
-          if(downEvent.target.tagName.toLowerCase() === "input" || downEvent.target.tagName.toLowerCase() === "select") return;
-          downEvent.preventDefault();
+        fieldset.ontouchstart = fieldset.onmousedown = (downEvent) => {
+          if((downEvent.target.parentElement.tagName.toLowerCase() === 'label' &&
+              downEvent.target.tagName.toLowerCase() !== 'input') ||
+              downEvent.target.tagName.toLowerCase() === 'label') {
+          // downEvent.preventDefault();
```
Comment (prushforth) [inline src/mapml/layers/MapMLLayer.js:691]: delete, probably?

**Review by prushforth — APPROVED** [review]
A few small edits required but good job fixing this problem! Thanks!

File: src/mapml/layers/MapMLLayer.js (hunk 6 above)
Comment (yhy0217) [reply src/mapml/layers/MapMLLayer.js]: Yes I think it is definitely necessary but not urgent, I did not find other bugs/issues except the layer control ones (which I have fixed on this branch) during the debugging process! :)

File: src/mapml/control/LayerControl.js (hunk 1 above)
Comment (yhy0217) [reply src/mapml/control/LayerControl.js]: New knowledge! Thanks!

File: src/mapml/layers/MapMLLayer.js (hunk 4 above)
Comment (yhy0217) [reply src/mapml/layers/MapMLLayer.js]: I believe it is called chaining assignment (there is an example of it in mdn: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment). it really helpful in saving time and space!

Comment (prushforth) [reply src/mapml/layers/MapMLLayer.js]: New knowledge thanks!

**Pull request merged**